File: 백준/Silver/24445. 알고리즘 수업 － 너비 우선 탐색 2/알고리즘 수업 － 너비 우선 탐색 2.java
import java.io.*;
import java.util.*;

public class Main {
    public static int N,M,R;
    public static List<Integer> a[];
    public static boolean[] visited;
    public static int[] order;
    public static int count = 0;
    
    public static void bfs(int x) {
        Queue<Integer> q = new LinkedList<>();
        visited[x] = true;
        q.offer(x);
        
        while(!q.isEmpty()) {
            int num = q.poll();
            count++;
            order[num] = count;
            for(Integer i : a[num]) {
                if(!visited[i]) {
                    q.offer(i);
                    visited[i] = true;
                }
            }
        } 
    }
    
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine(), " ");
        int N = Integer.parseInt(st.nextToken());
        int M = Integer.parseInt(st.nextToken());
        int R = Integer.parseInt(st.nextToken());
        
        a = new ArrayList[N+1];
        visited = new boolean[N+1];
        order = new int[N+1];
        
        for(int i = 0; i <= N; i++) {
            a[i] = new ArrayList<>();
        }
        
        for(int i=0; i<M; i++) {
            st = new StringTokenizer(br.readLine(), " ");
            int U = Integer.parseInt(st.nextToken());
            int V = Integer.parseInt(st.nextToken());
            a[U].add(V);
            a[V].add(U);
        }
        
        for(int i = 1; i <= N; i++) {
            Collections.sort(a[i], Collections.reverseOrder());
        }
        
        bfs(R);

        for (int i = 1; i <= N; i++) {
            System.out.println(order[i]);
        }
    }
}

File: 백준/Silver/24445. 알고리즘 수업 － 너비 우선 탐색 2/알고리즘 수업 － 너비 우선 탐색 2.py
import sys
from collections import deque
sys.setrecursionlimit(10**6)

def bfs(graph, R, N):
    visited = [False] * (N + 1)
    order = [0] * (N + 1)
    visited[R] = True
    cnt = 1
    order[R] = cnt
    queue = deque([R])
    
    while queue:
        current = queue.popleft()
        
        for next_node in sorted(graph[current], reverse=True):
            if not visited[next_node]:
                cnt += 1
                visited[next_node] = True
                order[next_node] = cnt
                queue.append(next_node)
                
    return order

N, M, R = map(int, sys.stdin.readline().split())
graph = [[] for _ in range(N + 1)]

for _ in range(M):
    u, v = map(int, sys.stdin.readline().split())
    graph[u].append(v)
    graph[v].append(u)
    
result = bfs(graph, R, N)

for i in range(1, N + 1):
    print(result[i])